### PR TITLE
Handling native exeptions

### PR DESCRIPTION
**Pull request by Rsedaikin**

### CODE REVIEW

File: skiko/build.gradle.kts [hunk 1]
```diff
@@ -268,7 +268,7 @@ tasks.withType(CppCompile::class.java).configureEach {
             compilerArgs.addAll(
                 listOf(
                     "-fno-rtti",
-                    "-fno-exceptions",
+                    "-fexceptions",
```
Comment (olonho) [inline skiko/build.gradle.kts:271]: Do not think it's a good idea, let's avoid using exceptions.

File: skiko/src/jvmMain/cpp/common/DirectContext.cc [hunk 2]
```diff
@@ -41,8 +43,12 @@ extern "C" JNIEXPORT jlong JNICALL Java_org_jetbrains_skia_DirectContext__1nMake
 
 extern "C" JNIEXPORT void JNICALL Java_org_jetbrains_skia_DirectContext__1nFlush
   (JNIEnv* env, jclass jclass, jlong ptr) {
-    GrDirectContext* context = reinterpret_cast<GrDirectContext*>(static_cast<uintptr_t>(ptr));
-    context->flush(GrFlushInfo());
+    try {
+      GrDirectContext* context = reinterpret_cast<GrDirectContext*>(static_cast<uintptr_t>(ptr));
```
Comment (olonho) [inline skiko/src/jvmMain/cpp/common/DirectContext.cc:47]: Does Skia throws exceptions?

File: skiko/src/jvmMain/cpp/windows/directXRedrawer.cc [hunk 3]
```diff
@@ -326,33 +324,37 @@ extern "C"
...
-
-        RECT windowRect;
-        GetWindowRect(d3dDevice->window, &windowRect);
-        unsigned int w = windowRect.right - windowRect.left;
-        unsigned int h = windowRect.bottom - windowRect.top;
-        GR_D3D_CALL_ERRCHECK(d3dDevice->swapChain->ResizeBuffers(BuffersCount, w, h, DXGI_FORMAT_R8G8B8A8_UNORM, 0));
-
-        swapChainFactory.reset(nullptr);
-        swapChainFactory.reset(nullptr);
+        try {
```
Comment (olonho) [inline skiko/src/jvmMain/cpp/windows/directXRedrawer.cc:327]: Not sure if it's a good idea. If DirectX indeed throws exceptions - we shall use __try to catch system-level exceptions.

Comment (Rsedaikin) [reply skiko/src/jvmMain/cpp/windows/directXRedrawer.cc:327]: I am not sure about using `__try __except` block instead standart `try catch`, but a lot of developers says to not using it unless you defenetly need it and I cannot say that only DirectX throw exceptions.

File: skiko/src/jvmMain/cpp/common/DirectContext.cc (hunk 2 above)
Comment (Rsedaikin) [reply skiko/src/jvmMain/cpp/common/DirectContext.cc:47]: yes - https://youtrack.jetbrains.com/issue/TBX-6557 (see attached log file)

File: skiko/build.gradle.kts (hunk 1 above)
Comment (Rsedaikin) [reply skiko/build.gradle.kts:271]: without it the linux version won't compile

Comment (olonho) [reply skiko/build.gradle.kts:271]: We shalln't try/catch in non-Windows code

File: skiko/src/jvmMain/cpp/common/jni_helpers.cc [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+    }
+
+    jstring propertyName= env->NewStringUTF("skiko.win.renderexceptionhandling.enabled");
+    jstring propertyString = (jstring)env->CallStaticObjectMethod(systemClass, getPropertyMethod, propertyName);
+    if (propertyString == 0)
+    {
+        return false;
+    }
+    const char *property = env->GetStringUTFChars(propertyString, 0);
+    bool result = !strcmp(property, "true");
```
Comment (olonho) [inline skiko/src/jvmMain/cpp/common/jni_helpers.cc:28]: `strncmp(property, "true", 4)` would be safer.

File: skiko/src/jvmMain/cpp/common/impl/Managed.cc [hunk 5]
```diff
@@ -5,7 +5,7 @@ typedef void (*FreeFunction)(void*);
 
 extern "C" JNIEXPORT void JNICALL Java_org_jetbrains_skia_impl_Managed__1nInvokeFinalizer
   (JNIEnv* env, jclass jclass, jlong finalizerPtr, jlong ptr) {
-    void* instance = reinterpret_cast<void*>(static_cast<uintptr_t>(ptr));
-    FreeFunction finalizer = reinterpret_cast<FreeFunction>(static_cast<uintptr_t>(finalizerPtr));
-    finalizer(instance);
+      void* instance = reinterpret_cast<void*>(static_cast<uintptr_t>(ptr));
```
Comment (olonho) [inline skiko/src/jvmMain/cpp/common/impl/Managed.cc:8]: Why this change?

File: skiko/src/jvmMain/cpp/linux/HardwareInfo.cc [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+        }
+        return result;
+    }
+
+    JNIEXPORT jstring JNICALL Java_org_jetbrains_skiko_RenderExceptionsHandlerKt_getNativeGraphicsAdapterInfo(
+        JNIEnv *env, jobject object)
+    {
+        std::stringstream stream;
+        std::string searchSentence = "VGA compatible controller: ";
+        stream << "lspci | grep " << "\"" << searchSentence << "\"";
```
Comment (olonho) [inline skiko/src/jvmMain/cpp/linux/HardwareInfo.cc:30]: This looks somewhat hacky and shalln't belong to Skiko generic code, IMO.

Comment (Rsedaikin) [reply skiko/src/jvmMain/cpp/linux/HardwareInfo.cc]: Done.

File: skiko/src/jvmMain/cpp/common/impl/Managed.cc (hunk 5 above)
Comment (Rsedaikin) [reply skiko/src/jvmMain/cpp/common/impl/Managed.cc]: Done.

File: skiko/src/jvmMain/cpp/common/jni_helpers.cc (hunk 4 above)
Comment (Rsedaikin) [reply skiko/src/jvmMain/cpp/common/jni_helpers.cc]: Done.

File: skiko/src/jvmMain/cpp/windows/directXRedrawer.cc (hunk 3 above)
Comment (Rsedaikin) [reply skiko/src/jvmMain/cpp/windows/directXRedrawer.cc]: Done.

File: skiko/src/jvmMain/cpp/common/DirectContext.cc (hunk 2 above)
Comment (Rsedaikin) [reply skiko/src/jvmMain/cpp/common/DirectContext.cc]: Done.

File: skiko/build.gradle.kts (hunk 1 above)
Comment (Rsedaikin) [reply skiko/build.gradle.kts]: Done.

File: skiko/src/jvmMain/cpp/common/jni_helpers.cc [hunk 7]
```diff
@@ -0,0 +1,110 @@
+#if SK_BUILD_FOR_WIN
+
+#include "jni_helpers.h"
+
+JavaVM *jvm = NULL;
```
Comment (olonho) [inline skiko/src/jvmMain/cpp/common/jni_helpers.cc:5]: static

Comment (Rsedaikin) [reply skiko/src/jvmMain/cpp/common/jni_helpers.cc]: Done.

File: samples/SkiaJvmSample/build.gradle.kts [hunk 8]
```diff
@@ -53,6 +53,8 @@ val casualRun = tasks.named<JavaExec>("run") {
     systemProperty("skiko.fps.enabled", "true")
     systemProperty("skiko.linux.autodpi", "true")
     systemProperty("skiko.hardwareInfo.enabled", "true")
+    systemProperty("skiko.win.renderexceptionlogger.enabled", "true")
```
Comment (olonho) [inline samples/SkiaJvmSample/build.gradle.kts:56]: renderexceptionlogger -> exception.logger
renderexceptionhandling -> exception.handler

File: skiko/buildSrc/src/main/kotlin/properties.kt [hunk 9]
```diff
@@ -81,6 +81,7 @@ class SkikoProperties(private val myProject: Project) {
     val deployVersion: String
         get() {
             val version = myProject.property("deploy.version") as String
+            File("src/jvmMain/resources/version").writeText(version)
```
Comment (olonho) [inline skiko/buildSrc/src/main/kotlin/properties.kt:84]: It doesn't belong to getter.

File: skiko/src/jvmMain/cpp/common/jni_helpers.cc [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+    {
+        return false;
+    }
+    const char *property = env->GetStringUTFChars(propertyString, 0);
+    bool result = !strncmp(property, "true", 4);
+    env->ReleaseStringUTFChars(propertyString, property);
+    return result;
+}
+
+std::string getDescription(DWORD code)
```
Comment (olonho) [inline skiko/src/jvmMain/cpp/common/jni_helpers.cc:33]: `const char* getDescription(DWORD code) {`

File: skiko/src/jvmMain/cpp/common/jni_helpers.cc [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+        return "EXCEPTION_STACK_OVERFLOW";
+    default:
+        return "UNKNOWN EXCEPTION";
+    }
+}
+
+void logJavaException(JNIEnv *env, const char *function, DWORD sehCode)
+{
+    std::ostringstream oss;
+    oss << "Native exception in [" << function << "]:" << std::endl;
```
Comment (olonho) [inline skiko/src/jvmMain/cpp/common/jni_helpers.cc:85]: Why C++ streams? Maybe just regular `snprintf()` to fixed size buffer.

File: skiko/src/jvmMain/cpp/include/jni_helpers.h [hunk 12]
```diff
@@ -1,7 +1,15 @@
 #pragma once
 
+#include <jni.h>
+
+#if SK_BUILD_FOR_WIN
+#include <windows.h>
+#include <sstream>
+void logJavaException(JNIEnv *env, const char * function, DWORD sehCode);
```
Comment (olonho) [inline skiko/src/jvmMain/cpp/include/jni_helpers.h:8]: Why put it here at all?

File: skiko/src/jvmMain/cpp/windows/HardwareInfo.cc [hunk 13]
```diff
@@ -0,0 +1,64 @@
+#ifdef SK_DIRECT3D
+
+#include <locale>
```
Comment (olonho) [inline skiko/src/jvmMain/cpp/windows/HardwareInfo.cc:3]: Why locale and streams needed here?

File: skiko/src/jvmMain/cpp/windows/directXRedrawer.cc [hunk 14]
```diff
@@ -439,16 +447,22 @@ extern "C"
...
     {
-        DirectXDevice *d3dDevice = fromJavaPointer<DirectXDevice *>(devicePtr);
-        const UINT64 fenceValue = d3dDevice->fenceValues[d3dDevice->bufferIndex];
-        d3dDevice->bufferIndex = d3dDevice->swapChain->GetCurrentBackBufferIndex();
-        if (d3dDevice->fence->GetCompletedValue() < fenceValue)
-        {
-            GR_D3D_CALL_ERRCHECK(d3dDevice->fence->SetEventOnCompletion(fenceValue, d3dDevice->fenceEvent));
-            WaitForSingleObjectEx(d3dDevice->fenceEvent, INFINITE, FALSE);
+        __try {
+            DirectXDevice *d3dDevice = fromJavaPointer<DirectXDevice *>(devicePtr);
```
Comment (olonho) [inline skiko/src/jvmMain/cpp/windows/directXRedrawer.cc:451]: We need to test it, code is pretty fragile.

File: skiko/src/jvmMain/kotlin/org/jetbrains/skiko/Library.kt [hunk 15]
```diff
@@ -98,6 +98,10 @@ object Library {
             t.printStackTrace()
         }
     }
+
+    val skikoVersion: String by lazy {
+        Library::class.java.getResource("/version").readText()
```
Comment (olonho) [inline skiko/src/jvmMain/kotlin/org/jetbrains/skiko/Library.kt:103]: Please use generated source file here instead.

File: skiko/src/jvmMain/kotlin/org/jetbrains/skiko/redrawer/Direct3DRedrawer.kt [hunk 16]
```diff
@@ -17,7 +17,7 @@ internal class Direct3DRedrawer(
 
     private val device = createDirectXDevice(getAdapterPriority(), layer.contentHandle).also {
         if (it == 0L || !isVideoCardSupported(layer.renderApi)) {
-            throw IllegalArgumentException("Failed to create DirectX12 device.")
+            throw Exception("Failed to create DirectX12 device.")
```
Comment (olonho) [inline skiko/src/jvmMain/kotlin/org/jetbrains/skiko/redrawer/Direct3DRedrawer.kt:20]: IllegalStateException

File: skiko/src/jvmMain/cpp/windows/directXRedrawer.cc (hunk 14 above)
Comment (Rsedaikin) [reply skiko/src/jvmMain/cpp/windows/directXRedrawer.cc:452]: I think we should try to keep this code, and maybe we can collect some useful information about the problem if SEH occurs in this part of the code.
PS. In this part of code, I am unable to simulate SEH on my PC, but on my old laptop (with Intel HD 520), I sometimes encountered exceptions when resizing.

File: skiko/src/jvmMain/kotlin/org/jetbrains/skiko/Library.kt (hunk 15 above)
Comment (Rsedaikin) [reply skiko/src/jvmMain/kotlin/org/jetbrains/skiko/Library.kt]: Done.

File: skiko/src/jvmMain/kotlin/org/jetbrains/skiko/redrawer/Direct3DRedrawer.kt (hunk 16 above)
Comment (Rsedaikin) [reply skiko/src/jvmMain/kotlin/org/jetbrains/skiko/redrawer/Direct3DRedrawer.kt]: Done.

File: samples/SkiaJvmSample/build.gradle.kts (hunk 8 above)
Comment (Rsedaikin) [reply samples/SkiaJvmSample/build.gradle.kts]: Done.

File: skiko/buildSrc/src/main/kotlin/properties.kt (hunk 9 above)
Comment (Rsedaikin) [reply skiko/buildSrc/src/main/kotlin/properties.kt]: Done.

File: skiko/src/jvmMain/cpp/common/jni_helpers.cc (hunk 10 above)
Comment (Rsedaikin) [reply skiko/src/jvmMain/cpp/common/jni_helpers.cc]: Done.

File: skiko/src/jvmMain/cpp/common/jni_helpers.cc (hunk 11 above)
Comment (Rsedaikin) [reply skiko/src/jvmMain/cpp/common/jni_helpers.cc]: Why not? It's convenient to use, but as you want I change to `snprintf()`. Done.

File: skiko/src/jvmMain/cpp/include/jni_helpers.h (hunk 12 above)
Comment (Rsedaikin) [reply skiko/src/jvmMain/cpp/include/jni_helpers.h]: Done.

File: skiko/src/jvmMain/cpp/windows/HardwareInfo.cc (hunk 13 above)
Comment (Rsedaikin) [reply skiko/src/jvmMain/cpp/windows/HardwareInfo.cc]: Done.

File: skiko/build.gradle.kts [hunk 17]
```diff
@@ -34,6 +34,15 @@ repositories {
     mavenCentral()
 }
 
+fun putLibraryVersion() {
```
Comment (olonho) [inline skiko/build.gradle.kts:37]: Cool

File: skiko/build.gradle.kts [hunk 18]
```diff
@@ -34,6 +34,15 @@ repositories {
...
+    val versionFile = File("src/jvmMain/kotlin/org/jetbrains/skiko/Version.kt")
```
Comment (olonho) [inline skiko/build.gradle.kts:38]: Better use `File(project.projectDir, "src/jvmMain/kotlin/org/jetbrains/skiko/Version.kt")`

File: skiko/build.gradle.kts [hunk 19]
```diff
@@ -34,6 +34,15 @@ repositories {
...
+    val outputBuilder = StringBuilder().apply {
```
Comment (olonho) [inline skiko/build.gradle.kts:39]: Multiline literals would be better here, i.e.
```
versionFile..writeText("""
            package org.jetbrains.skiko
            val skikoVersion = "${skiko.deployVersion}"
        """.trimIndent())
```

File: skiko/build.gradle.kts [hunk 20]
```diff
@@ -582,6 +591,8 @@ library {
     targetMachines.addAll(listOf(machines.macOS.x86_64, machines.linux.x86_64, machines.windows.x86_64))
     baseName.set("skiko-$target")
 
+    putLibraryVersion()
```
Comment (olonho) [inline skiko/build.gradle.kts:594]: It makes incorrect Gradle dependency graph and just doesn't belong here. I can make version support in separate PR.

**Review by olonho — APPROVED** [review]
